python/ccf/read_ledger.py

```diff
@@ -14,18 +14,51 @@ def indent(n):
     return " " * n
 
 
-def stringify_bytes(bs):
-    s = bs.decode()
-    if s.isprintable():
-        return s
-    if len(bs) > 0 and len(bs) <= 8:
-        n = int.from_bytes(bs, byteorder="little")
-        return f"<u{8 * len(bs)}: {n}>"
-    return bs
+def fmt_unsigned_int_little(data):
+    return (
+        f'<u{8 * len(data)}: {int.from_bytes(data, byteorder="little", signed=False)}>'
+    )
+
+
+def fmt_hex(data):
+    return data.hex()
+
+
+def fmt_str(data):
+    return data.decode()
+
+
+def fmt_json(data):
+    return json.dumps(json.loads(data), indent=2)
+
 
+# Dictionary of store table name regex to format function (first match is used)
+# Callers can specify additional rules (e.g. for application-specific
+# public tables) which get looked up first.
+default_format_rule = {"key": fmt_hex, "value": fmt_hex}
+default_tables_format_rules = {
+    "^public:ccf\.internal\..*$": {
+        "key": fmt_unsigned_int_little,
+        "value": fmt_json,
+    },
+    "^public:ccf\.gov\.(service|network|constitution).*$": {
+        "key": fmt_unsigned_int_little,
+        "value": fmt_json,
+    },
+    "^public:ccf\.gov\..*$": {"key": fmt_str, "value": fmt_json},
+    ".*": {"key": fmt_hex, "value": fmt_hex},
+}
 
-def print_key(indent_s, k, is_removed=False):
-    k = stringify_bytes(k)
+
+def find_rule(tables_format_rules, target_table_name):
+    for table_name_re, format_rules in tables_format_rules.items():
+        if re.compile(table_name_re).match(target_table_name):
```

Comment:
Not critical, since we don't have any performance bounds for this, but recompiling the regex here every time will get very slow for large ledgers/many rules. Should be possible to precompile (convert a `list-of-pairs-of-(string-and-rule)` to a `list-of-pairs-of-(regex-and-rule)`, and pass that here.